### PR TITLE
Clarify emitting behavior of GPU particles

### PR DESCRIPTION
**Pull request by Gatada**
Clarified the finished signal, as it is incorrect to say that the signal is issued once processing of all active particles have finished (at least according to the GitHub thread that explained this to me).

This time, modifying the XML file.

### CODE REVIEW

File: doc/classes/GPUParticles3D.xml
Comment (clayjohn) [inline doc/classes/GPUParticles3D.xml:153]: > Emitted when all active particles are no longer visible

I don't think this is accurate. Visibility has nothing to do with this signal. You can make your node invisible and it won't fire the signal.

Comment (Gatada) [reply doc/classes/GPUParticles3D.xml:153]: You say "visibility has nothing to do with this signal" - and that "make your node invisible and won't fire"? Sounds like visibility does have something to do with the signal?

**Comment by Gatada** [comment]
Here is where I am explained that the particle emitter will emit the ```finished``` signal once the particles are no longer visible: https://github.com/godotengine/godot/pull/76859

What I see in my test project concurs: the finished signal is received before the emitter is ready to emit, yet there are no particles visible on the screen.

**Comment by HolonProduction** [comment]
> Here is where I am explained that the particle emitter will emit the `finished` signal once the particles are no longer visible: #76859

What I meant when I said that, is that I can guarantee you that the particles will no longer be visible. But that doesn't mean that they are necessarily still processed.

**Comment by Gatada** [comment]
> > Here is where I am explained that the particle emitter will emit the `finished` signal once the particles are no longer visible: #76859
> 
> What I meant when I said that, is that I can guarantee you that the particles will no longer be visible. But that doesn't mean that they are necessarily still processed.

But that is, in my opinion, not relevant for the game developer. The only thing that is relevant is that when the ```finished``` signal is received, the game developer can call ```restart()``` with no risk of popping particles off screen. In other words, when finished is received, the particles may or may not be processing, but are either way no longer visible.

This is why I suggest the wording in my PR.

**Comment by HolonProduction** [comment]
To clarify: I am in favor of adding clarification about this to the docs.

But we need to make a difference between intended behavior and the current situation.

We should not change the part that says `Emitted when all active particles have finished processing` IMO. It is the _intended_ behavior and we can't remove the intended behavior from the docs. (It's especially important here, to keep consistent with the CPU particle nodes.)

That the node thinks it is processing for longer than the particles are visible does not sound like intended and defined behavior on the other hand. It's a quirk in the current code which would better be placed inside a **Note:**.

**Comment by HolonProduction** [comment]
Also when we find a suited modification for the docs. You should copy it over to the 2D GPU particles as well.

File: doc/classes/GPUParticles3D.xml
Comment (HolonProduction) [reply doc/classes/GPUParticles3D.xml:153]: The problem here is, that `visibility` already has a rather clear meaning in the godot context. (The property on nodes).

"Are no longer drawn" / "Are no longer displayed" would convey the same idea without using the word `visibility`.

Comment (clayjohn) [reply doc/classes/GPUParticles3D.xml:153]: I'm saying that even if you make you node invisible (and thus make all the particles no longer visible) the signal won't fire. Precisely because the signal is not connected to visibility whatsoever. 

The signal fires after all the nodes have finished processing (regardless of whether the particles are visible or not). But, as the current iteration of the docs points out, it's inexact due to timing issues between the CPU and the GPU.

Again, visibility does not play a factor.

**Comment by Gatada** [comment]
> To clarify: I am in favor of adding clarification about this to the docs.
> 
> But we need to make a difference between intended behavior and the current situation.
> 
> We should not change the part that says `Emitted when all active particles have finished processing` IMO. It is the _intended_ behavior and we can't remove the intended behavior from the docs. (It's especially important here, to keep consistent with the CPU particle nodes.)

Well, I thought we were documenting how things work, and not how we would like it to work? I would assume a future release that changes this, will be responsible for updating the docs? Surely this is how it should be?

> That the node thinks it is processing for longer than the particles are visible does not sound like intended and defined behavior on the other hand. It's a quirk in the current code which would better be placed inside a **Note:**.

I just want the docs to be actual helpful. As it is written, I wasted hours messaging to understand why the node was still ```emitting``` (i.e. true) even after having received ```finished```.

**Comment by HolonProduction** [comment]
> Well, I thought we were documenting how things work, and not how we would like it to work? I would assume a future release that changes this, will be responsible for updating the docs? Surely this is how it should be?

If a future release were to change things it would change nothing about the implementation of this signal. The signal is working as it should be. So I see no ground for changing it's behavior description based on, what I would call, undefined behavior at an other point.

**Side Note:**
I am no core dev so arguing with me about the philosophy of writing docs won't help the situation. It's my _personal_ opinion on how to write docs that help both developers and users.

> I just want the docs to be actual helpful.

:+1: 

> As it is written, I wasted hours messaging to understand why the node was still `emitting` (i.e. true) even after having received `finished`.

So first of all `emitting` is `false` when the signal finished (except you set it yourself). That's just how it is implemented. (Put simple: Emitting will be set to `false` after half of the `lifetime`)

The thing that needs clarification is that setting `emitting` to `true` won't guarantee the system to start even when `emitting` was `false` before. This fact has little to do with this signal. It was true in `4.1` as well.

`emitting` is just not the right way to start a one shot particle system. And something like this could go into the docs (preferably for the `emitting` properties description).

**Review by Gatada — COMMENTED** [review]
I believe this corrects my earlier misunderstandings, and clarifies things—while helping developers be more efficient.

File: doc/classes/GPUParticles3D.xml
Comment (HolonProduction) [inline doc/classes/GPUParticles3D.xml:153]: I'm not quite sure what "the emitter" should mean in this context.
```suggestion
				Emitted when all active particles have finished processing. To immediately emit new particles, call [method restart].
```

File: doc/classes/GPUParticles3D.xml
Comment (HolonProduction) [inline doc/classes/GPUParticles3D.xml:94]: Isn't that exactly what's not working?

**Review by HolonProduction — CHANGES_REQUESTED** [review]

File: doc/classes/GPUParticles3D.xml
Comment (Gatada) [reply doc/classes/GPUParticles3D.xml:154]: OK, but if that is what the note meant, then it has to be rephrased. As the current note in no way helps the game developer avoid or fix the issue of a delayed signal. I assume the delay will not occur by removing randomness in the emitter?

Comment (HolonProduction) [reply doc/classes/GPUParticles3D.xml:154]: > I assume the delay will not occur by removing randomness in the emitter?

As far as I know yes.

File: doc/classes/GPUParticles3D.xml
Comment (Gatada) [reply doc/classes/GPUParticles3D.xml:94]: ha ha.. Right, so this is where the note belongs. Adding it.

File: doc/classes/GPUParticles2D.xml
Comment (AThousandShips) [inline doc/classes/GPUParticles2D.xml:133]:
```suggestion
				[b]Note for one shot emitters:[/b] Due to the particles being computed on the GPU, there may be a short period after receiving the signal during which setting [member emitting] to [code]true[/code] will not restart the emission cycle. This delay is avoided by calling [method restart].
```
Did you mean this property? Same below

File: doc/classes/GPUParticles3D.xml
Comment (AThousandShips) [inline doc/classes/GPUParticles3D.xml:156]:
```suggestion
				[b]Note for one shot emitters:[/b] Due to the particles being computed on the GPU, there may be a short period after receiving the signal during which setting [member emitting] to [code]true[/code] will not spawn new particles. This delay is avoided by calling [method restart].
```
Here as well

File: doc/classes/GPUParticles2D.xml
Comment (HolonProduction) [inline doc/classes/GPUParticles2D.xml:65]: You will have to use `[b]Note:[/b]` to get the documentation to render the blue note boxes. Try to integrate "for one shot emitters" into the sentence after it.

File: doc/classes/GPUParticles3D.xml
Comment (Gatada) [reply doc/classes/GPUParticles3D.xml]: I believe this is resolved. Thanks for the help!

File: doc/classes/GPUParticles3D.xml
Comment (Gatada) [reply doc/classes/GPUParticles3D.xml]: Done, thanks!

File: doc/classes/GPUParticles2D.xml
Comment (Gatada) [reply doc/classes/GPUParticles2D.xml]: Corrected. Thanks!

**Comment by Gatada** [comment]
I believe the edit is now finally covering all feedback.

Sorry for taking up so much of your time. This has been a great learning experience, and I am very grateful for all comments! Thank you so much to everyone that took part!

If Godot is developed with the same level of care as this doc-change, then that makes me very happy. Godot will keep getting better, and I'll do everything I can to help out. Albeit more efficiently next time.

Any more comments or feedback?

**Comment by Gatada** [comment]
Now all it needs is a review. Do I ask for the review somewhere?

**Review by Gatada — COMMENTED** [review]
I think this should be pretty straight forward to merge? And adding this info, will make it easier for users.

**Comment by YuriSizov** [comment]
Still needs an approving review, but also you need to squash commits together into one. You can wait for an approval to make sure you don't need to make further edits, but you will need to do this eventually before we can merge.

Make sure that the final commit has a short but descriptive message (the title of this PR is a good option). See [this documentation](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), if you need help with squashing.

**Comment by Gatada** [comment]
OK, so I squashed and give the commit a succinct description.

**Review by mhilbrunner — APPROVED** [review]

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by Gatada** [comment]
Thanks, though far from my first. I had one in Oct 2022 as well ;-)

**Comment by akien-mga** [comment]
My bad! Might have been from a different GitHub account as GitHub doesn't seem to remember it: https://github.com/godotengine/godot/commits?author=Gatada

**Comment by Gatada** [comment]
I am not sure how these repos work, but here it is:

https://github.com/godotengine/godot-docs/commits/master?author=gatada

**Pull request merged**
